Move ReportBody and the remaining subtypes over to the new IPC serialization format

gavin-apple opened the pull request:
#### 9b5d93f7c707c1455f167db93fed6b0db34d5b20
<pre>
Move ReportBody and the remaining subtypes over to the new IPC serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=252563">https://bugs.webkit.org/show_bug.cgi?id=252563</a>
rdar://105677731

Reviewed by Alex Christensen.

This changes moves ReportBody and the remaining subtypes over to the new
IPC serialization format. This includes:
    - ReportBody
    - ViolationReportType
    - COEPDisposition
    - COEPInheritenceViolationReportBody
    - CORPViolationReportBody

* Source/WebCore/Modules/reporting/Report.cpp:
(WebCore::Report::create):
(WebCore::Report::asReportBodyType const):
* Source/WebCore/Modules/reporting/Report.h:
* Source/WebCore/Modules/reporting/ViolationReportType.h:
* Source/WebCore/loader/COEPInheritenceViolationReportBody.h:
(WebCore::COEPInheritenceViolationReportBody::encode const): Deleted.
(WebCore::COEPInheritenceViolationReportBody::decode): Deleted.
* Source/WebCore/loader/CORPViolationReportBody.h:
(WebCore::CORPViolationReportBody::encode const): Deleted.
(WebCore::CORPViolationReportBody::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;RefPtr&lt;WebCore::ReportBody&gt;&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;RefPtr&lt;WebCore::ReportBody&gt;&gt;::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/260642@main">https://commits.webkit.org/260642@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/762271d7472ff7f0eae187284a8da53d28fdeae2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108894 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17999 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41737 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/438 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118213 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112775 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/19456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9281 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101139 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114664 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/19456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97813 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42715 "Found 2 new test failures: webgl/2.0.y/conformance/misc/invalid-passed-params.html, webgl/2.0.y/conformance/programs/program-test.html (failure)") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/19456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29454 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84454 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10772 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30803 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11527 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7741 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16917 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50399 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7355 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13115 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/Modules/reporting/Report.h:
```diff
@@ -25,25 +25,33 @@
 
 class FormData;
 
+using ReportBodyType = std::variant<Ref<CSPViolationReportBody>, Ref<COEPInheritenceViolationReportBody>, Ref<CORPViolationReportBody>, Ref<DeprecationReportBody>, Ref<TestReportBody>>;
```
This shouldn't be changed in C++, but something like this will be needed in SerializedTypeInfo.cpp.  Do something like WebCore::TimingFunction, then in generate_serialized_type_info pretend ReportBody contains a variant of its subclasses.  We currently just skip it with the continue after "if type.members_are_subclasses:"

achristensen07 replied:
It looks like we can change it from a RefPtr to a Ref, though.

gavin-apple replied:
Thanks. I didn't know about "subclasses". Makes it wayyy cleaner.

What do you think can change from RefPtr to Ref? Report::m_body?

achristensen07 replied:
probably, but can be done in a separate PR.

gavin-apple committed:
Move ReportBody and the remaining subtypes over to the new IPC serialization format
https://bugs.webkit.org/show_bug.cgi?id=252563
rdar://105677731

Reviewed by Alex Christensen.

This changes moves ReportBody and the remaining subtypes over to the new
IPC serialization format. This includes:
    - ReportBody
    - ViolationReportType
    - COEPDisposition
    - COEPInheritenceViolationReportBody
    - CORPViolationReportBody

* Source/WebCore/Modules/reporting/Report.cpp:
(WebCore::Report::create):
(WebCore::Report::asReportBodyType const):
* Source/WebCore/Modules/reporting/Report.h:
* Source/WebCore/Modules/reporting/ViolationReportType.h:
* Source/WebCore/loader/COEPInheritenceViolationReportBody.h:
(WebCore::COEPInheritenceViolationReportBody::encode const): Deleted.
(WebCore::COEPInheritenceViolationReportBody::decode): Deleted.
* Source/WebCore/loader/CORPViolationReportBody.h:
(WebCore::CORPViolationReportBody::encode const): Deleted.
(WebCore::CORPViolationReportBody::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<RefPtr<WebCore::ReportBody>>::encode): Deleted.
(IPC::ArgumentCoder<RefPtr<WebCore::ReportBody>>::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/260642@main
